src/uwtools/tests/test_cli.py

```diff
@@ -130,77 +131,82 @@ def test_dispatch_config_realize():
-    assert cli.dispatch_config_translate(args) is False
+    assert cli._dispatch_config_translate(args) is False
 
 
-def test_dispatch_config_validate_yaml():
-    args = ns(input_file=1, input_format="yaml", schema_file=3)
+def test__dispatch_config_validate_yaml():
+    args = ns(input_file=1, input_format=FORMAT.yaml, schema_file=3)
     with patch.object(cli.uwtools.config.validator, "validate_yaml") as m:
-        cli.dispatch_config_validate(args)
+        cli._dispatch_config_validate(args)
     assert m.called_once_with(args)
 
 
 def test_dispath_config_validate_unsupported():
     args = ns(input_file=1, input_format="jpg", schema_file=3)
-    assert cli.dispatch_config_validate(args) is False
+    assert cli._dispatch_config_validate(args) is False
 
 
-@pytest.mark.parametrize("params", [("run", "dispatch_forecast_run")])
-def test_dispatch_forecast(params):
+@pytest.mark.parametrize("params", [("run", "_dispatch_forecast_run")])
+def test__dispatch_forecast(params):
     submode, funcname = params
     args = ns(submode=submode)
     with patch.object(cli, funcname) as m:
-        cli.dispatch_forecast(args)
+        cli._dispatch_forecast(args)
     assert m.called_once_with(args)
 
 
-def test_dispatch_forecast_run():
+def test__dispatch_forecast_run():
     args = ns(config_file=1, forecast_model="foo")
     with patch.object(cli.uwtools.drivers.forecast, "FooForecast", create=True) as m:
         CLASSES = {"foo": getattr(cli.uwtools.drivers.forecast, "FooForecast")}
         with patch.object(cli.uwtools.drivers.forecast, "CLASSES", new=CLASSES):
-            cli.dispatch_forecast_run(args)
+            cli._dispatch_forecast_run(args)
     assert m.called_once_with(args)
     m().run.assert_called_once_with()
 
 
-@pytest.mark.parametrize("params", [("render", "dispatch_template_render")])
-def test_dispatch_template(params):
+@pytest.mark.parametrize("params", [("render", "_dispatch_template_render")])
+def test__dispatch_template(params):
     submode, funcname = params
     args = ns(submode=submode)
     with patch.object(cli, funcname) as m:
-        cli.dispatch_template(args)
+        cli._dispatch_template(args)
     assert m.called_once_with(args)
 
 
-def test_dispatch_template_render_yaml(caplog):
+def test__dispatch_template_render_yaml(caplog):
     logging.getLogger().setLevel(logging.DEBUG)
     args = ns(
         input_file=1,
         output_file=2,
         values_file=3,
+        values_format=4,
         key_eq_val_pairs=["foo=88", "bar=99"],
-        values_needed=5,
-        dry_run=6,
+        values_needed=6,
+        dry_run=7,
     )
     with patch.object(cli.uwtools.config.templater, "render") as m:
         with patch.object(cli.sys, "argv", ["foo", "--bar", "88"]):
-            cli.dispatch_template_render(args)
+            cli._dispatch_template_render(args)
     assert m.called_once_with(args)
     assert logged(caplog, "Command: foo --bar 88")
```

Comment:
I can't say that I understand this test. We're patching `argv` and we want to make sure those values are being processed by argparse. But then we also check that `m` (outside its scope) is called (at least?) once with `args` though?

Reply:
Good eye. This test is weird because the function it tests is weird, because that function is doing something no other function does: In addition to calling into the right place in the `uwtools` library to access the requested functionality (`uwtools.config.templater.render()`) and passing arguments that would normally have been parsed from `argv`, it's also accessing `argv` directly to log the command that was issued from the command line. This is unfortunate, because nowhere else this deep into the CLI code -- and certainly not downstream of it, as we'd previously ensured -- is `argv` being accessed: Everywhere else, in all the `_dispath_*()` functions, `argv` has already been parsed and the functions are receiving only the parsed, not raw, `args`. But we had a request to log the command-line command for re-run capability, and that's why `argv` is being accessed here.

All the tests of these `_dispatch_*()` functions are pretty basic plubming tests, just making sure that the dispatch function has access to all the arguments it needs, and that it wires them up to the library code correctly. That's what's tested here with `args`, the mock `m` standing in for `uwtools.config.templater.render()`, and the `called_once_with(args)` assertion. Totally orthogonal to that is the mocking of `argv` and the assertion that the code logs what would have appeared on the command line.

I don't like it either.

One potential improvement would be to make two separate tests to avoid the illusion that the two things being tested are related in any way, because they aren't.

Another improvement might be to factor this logging of the original command line out and do it for _all_ commands, which would make this one dispatch function less of a snowflake.

What do you think?

Reply:
I'm going ahead with moving the command logging to the `main()` function (it's only active in `--verbose` mode) so that it applies to all commands. We can change later if needed, but for now this makes the code and tests more consistent.